url708d477f92_zSakare_SENG3011_tgz-pJ8-test_runner_TestRunnerJ8/src/main/generator/TradeSignalGenerator.java

```diff
@@ -0,0 +1,380 @@
+				// Check if the our date is after the ask list dates.
+				if (date.after(ask.getDateTime())) { 
+					if (lowest > ask.getPrice()) { 
+						lowest = ask.getPrice();
+					}
+				}
+			} catch (NullPointerException e) {
+				// Mark ask orders for clean up.
+				nullAsks.add(ask);
+			}
+		}
+		
+		// Clean up null orders.
+		askList.removeAll(nullAsks);
+		return lowest;
+	}
+	
+	public List<Order> newMomentumOrders(String volume, Integer lookBackPeriod, Double threshold) {
+		
+		List<Order> newOrders = new ArrayList<Order>();
+		
+		// Grab all the required order fields
+		String instrument = askList.get(FIRST_ELEMENT).getInstrument();
+		
+		// If user does not supply anything, use default values.
+		if (lookBackPeriod == null) {
+			lookBackPeriod = DEFAULT_LOOK_BACK_N;
+		}
+		if (threshold == null) {
+			threshold = DEFAULT_THRESHOLD;
+		}
+		
+		int maxIterations = tradeList.size()/lookBackPeriod;
+		boolean paired = false;
+		for (int j = 0; j < maxIterations; j++) {
+			// Look back on N number of trades to work out the momentum of the market.
+			double tradeReturn = 0.0;
+			Order finalTrade = null;
+			int shiftAcross = j*lookBackPeriod;
+			for (int i = shiftAcross; i < lookBackPeriod+shiftAcross && i < tradeList.size()-1; i++) {
+				Order previousTrade = tradeList.get(i); 
+				Order tradeToCompare = tradeList.get(i+1);
+				double calculatedReturn = ((tradeToCompare.getPrice() - previousTrade.getPrice())/previousTrade.getPrice());
+				tradeReturn += calculatedReturn;
+				finalTrade = tradeToCompare;
+				
+				// Calculate the current moving average.
+				double movingAverage = tradeReturn/i;
+				// Check if the moving average breaks the threshold.
+				if (movingAverage >= threshold && !paired) {
+					// Average broke positive threshold, indicates we should buy (stock getting better)
+					OrderBuilder orderBuilder = new OrderBuilderImpl(instrument,
+													finalTrade.getDateTime(),
+													findBidPriceAtTime(finalTrade.getDateTime()),
+													volume,
+													ORDER_BID,
+													ALGORITHMIC_BROKER_ID);
+					
+					Order newOrder = orderBuilder.build();
+					if (newOrder != null) {
+						printOrder(newOrder);
+						newOrders.add(newOrder);
+						paired = true;
+					}
+				} else if (movingAverage <= -threshold && paired) {
+					// Average broke negative threshold, indicates we should sell (stock getting worse)
+					OrderBuilder orderBuilder = new OrderBuilderImpl(instrument,
+													finalTrade.getDateTime(),
+													findAskPriceAtTime(finalTrade.getDateTime()),
+													volume,
+													ORDER_ASK,
+													ALGORITHMIC_BROKER_ID);
+					
+					Order newOrder = orderBuilder.build();
+					if (newOrder != null) {
+						printOrder(newOrder);
+						newOrders.add(newOrder);
+						paired = false;
+					}
+				}
+			}
+		}
+		
+		// Match any new orders.
+		if (newOrders.size() % 2 != 0) {
+			newOrders.remove(newOrders.size()-1);
+		}
+		return newOrders;
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 179:247, 252:320.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.